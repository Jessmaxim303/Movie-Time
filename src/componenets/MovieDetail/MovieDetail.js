import React, { Component } from 'react';
import './MovieDetail.scss';
import { connect } from 'react-redux';
import { postRating, deleteRating, getRatings, getMovies } from '../../actions';
import { postRatingToApi, deleteRatingFromApi, fetchRatingsAPI, fetchMoviesAPI } from '../../apiCalls/apiCalls';
import PropTypes from 'prop-types';

export class MovieDetail extends Component {
  constructor() {
    super();
    this.state = {
      currentRating: 0,
      visualClass: false
    }
  }

  handleRatingChange = e => {
    this.setState({[e.target.name]: parseInt(e.target.value)})
  }

  loadRatings() {
    fetchRatingsAPI(this.props.user.id)
    .then(data => {
      this.props.getRatings(data.ratings)
    })
  }

  loadMovies() {
    fetchMoviesAPI()
    .then(data => {
      this.props.getMovies(data.movies)
    })
  }

  submitRating() {
    let movieId = this.props.selectedMovie.id
    let newRating = this.state.currentRating
    let oldRating = this.props.ratings.find(rating => this.props.selectedMovie.id === rating.movie_id);
    let userId = this.props.user.id
    if (!oldRating) {
      postRatingToApi(movieId, newRating, userId)
        .then(res => {
          this.loadRatings()
          this.loadMovies()
        })
    } else {
      deleteRatingFromApi(oldRating.id, userId)
        .then(res => {
          postRatingToApi(movieId, newRating, userId)
            .then(res => {
              this.loadRatings()
              this.loadMovies()
            })
        })
    }
    this.setState({visualClass: true});
  }

  clearRating() {
    let oldRating = this.props.ratings.find(rating => this.props.selectedMovie.id === rating.movie_id);
    let userId = this.props.user.id
    deleteRatingFromApi(oldRating.id, userId)
      .then(res => {
        this.loadRatings()
        this.loadMovies()
      })
    this.setState({currentRating: 0})
  }

  componentDidMount() {
    let previousRating = this.props.ratings.find(rating => this.props.selectedMovie.id === rating.movie_id);
    if(previousRating) {
      this.setState({currentRating: previousRating.rating});
    }
  }

  render() {
    let {title, id, average_rating, poster_path, backdrop_path, release_date, overview} = this.props.selectedMovie
    const backgroundStyling = {
      background: `linear-gradient(to top, rgba(42, 42, 42, .75), rgba(42, 42, 42, 0)), url(${backdrop_path}) no-repeat center top`,
      backgroundSize: 'cover',
    }

    return (
      <section className='movie-detail' style={backgroundStyling} key={id}>
        <div className='movie-detail__info' >
          <img 
            className='movie-detail__poster' 
            alt={`movie poster for ${title}`} 
            src={poster_path} />
          <div className='movie-detail__details'>
            <h2 className='movie-detail__title'>{title}</h2>
            <p className='movie-detail__release-date'>
              {`Release Date: ${release_date}`}
            </p>
            <p className='movie-detail__average-rating'>
              {`Average Rating: ${average_rating.toFixed(2)}`}
            </p>
          {this.props.user ?
          <>
            <div className='movie-detail__user-rating'>
              <label for='currentRating'>User Rating: { this.state.currentRating || '-' }</label>
              <p className={this.state.visualClass ? 'movie-detail__rating-active' : 'movie-detail__rating-inactive'}>
                Thank you for rating!
              </p>
            </div>
            <input 
              className='movie-detail__rating-input' 
              onChange={this.handleRatingChange} 
              value={this.state.currentRating} 
              type='range' 
              min='1' 
              max='10' 
              id='user-rating' 
              name='currentRating' 
              value={this.state.currentRating} />
            <div className='movie-detail__rating-buttons'>
              <button
                className='movie-detail__button movie-detail__delete-rating'
                onClick={() => this.clearRating()}>
                  Delete Rating
              </button>
              <button 
                className='movie-detail__button movie-detail__add-rating' 
                onClick={() => this.submitRating()}>
                  Add Rating
              </button>
            </div>
          </>
          : null}
            <p className='movie-detail__overview'>{overview}</p>
          </div>
        </div>
      </section>
    )
  }
}

export const mapStateToProps = state => ({
  movies: state.movies,
  ratings: state.ratings,
  user: state.user
})

export const mapDispatchToProps = dispatch => ({
  postRating: (rating) => dispatch(postRating(rating)),
  deleteRating: (rating) => dispatch(deleteRating(rating)),
  getRatings: (ratings) => dispatch(getRatings(ratings)),
  getMovies: (movies) => dispatch(getMovies(movies))
})

export default connect(mapStateToProps, mapDispatchToProps)(MovieDetail);

MovieDetail.propTypes = {
  movies: PropTypes.array,
  ratings: PropTypes.array,
  user: PropTypes.object
};
