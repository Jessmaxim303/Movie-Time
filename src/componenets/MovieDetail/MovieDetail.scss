@import '../../styles/mixins.scss';
@import '../../styles/variables.scss';

.movie-detail {
  @include display-flex (row, center, center);
  flex: 1;
  &:first-child {
    flex: 1;
  }
  &__info {
    @include display-flex (row, center, space-around);
    border-radius: 0.75em;
    color: $text-color;
    font-size: 1.25rem;
    margin: 0.5em;
    padding: 0.75em;
    width: 70%;
    background-color: $card-background;
    box-shadow: $black-box-shadow
  }
  &__poster {
    border-radius: 0.75em;
    box-shadow: $black-box-shadow;
    margin: 0.5em 0;
    width: 40%;
  }
  &__details {
    @include display-flex (column, center, center);
    width: 45%;
  }
  &__title {
    font-family: $title-font;
    font-weight: 900;
    font-size: 2.5rem;
    margin: 0;
    text-align: center;
  }
  &__release-date {
    margin: 0.5em 0;
  }
  &__average-rating {
    font-weight: 700;
  }
  &__overview {
    height: 10em;
    overflow: scroll;
    text-align: left;
    width: 100%;
  }
  &__rating-input{
    margin: 1em 0 0.5em;
  }
  &__rating-buttons {
    @include display-flex(row nowrap, center, space-between);
  }
  &__button {
    @include button_style;
    margin: 0.5em;
    padding: 1em;
    width: 10em;
    &:hover {
      transform: none;
      background-color: darken($primary-color, 10);
    }
  }
  &__delete-rating {
    background-color: $accent-color;
  }
  &__user-rating {
    @include display-flex(row nowrap, center, space-between);
  }
  &__rating-inactive {
    display: none;
  }
  &__rating-active {
    display: block;
    color: lighten($primary-color, 10);
    animation: flash-rating 2s normal;
    margin: 0;
  }
}

@keyframes flash-rating {
  0% {
    color: lighten($primary-color, 10);
  }
  50% {
    color: $accent-color;
  }
  100% {
    color: lighten($primary-color, 10);
  }
}