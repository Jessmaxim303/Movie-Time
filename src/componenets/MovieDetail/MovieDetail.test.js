import React from 'react';
import { shallow } from 'enzyme';
import { MovieDetail, mapStateToProps, mapDispatchToProps } from './MovieDetail.js';
import { postRating } from '../../actions/index.js'


describe('MovieDeatil', () => {
  let wrapper, mockMovie;
  beforeEach(() => {
    mockMovie = {
      "id": 21,
      "title": "Sonic the Hedgehog",
      "poster_path": "https://image.tmdb.org/t/p/original//aQvJ5WPzZgYVDrxLX4R6cLJCEaQ.jpg",
      "backdrop_path": "https://image.tmdb.org/t/p/original//tCUcf3oNWMW8kwAj3WC6CvIN5ah.jpg",
      "release_date": "2020-02-12",
      "overview": "Based on the global blockbuster videogame franchise from Sega, Sonic the Hedgehog tells the story of the worldâ€™s speediest hedgehog as he embraces his new home on Earth. In this live-action adventure comedy, Sonic and his new best friend team up to defend the planet from the evil genius Dr. Robotnik and his plans for world domination.",
      "average_rating": 7
  }
    wrapper = shallow(<MovieDetail selectedMovie={mockMovie} />)
  })
  it('should match a snapshot', () => {
    expect(wrapper).toMatchSnapshot()
  })
  describe('mapStateToProps', () => {
    it('should return an object with the movies and ratings array', () => {
      const mockState = {
        movies: [{title: 'Parasite', rating: 9}, {title: '1917', rating: 8}],
        ratings: [{id: 25, rating: 9}, {id: 27, rating: 4}],
        user: {id: 123, name: 'me'},
        loadingStatus: true
      }
      const expected = {
        movies: [{title: 'Parasite', rating: 9}, {title: '1917', rating: 8}],
        ratings: [{id: 25, rating: 9}, {id: 27, rating: 4}],
        user: {id: 123, name: 'me'}
      }
      const mappedProps = mapStateToProps(mockState)
      expect(mappedProps).toEqual(expected)
    })
  })
  describe('mapDispatchToProps', () => {
    it('should call dispatch with the postRating action when postRating is invoked', () => {
      const mockDispatch = jest.fn();
      const rating = {id: 123, rating: 4};
      const action = postRating(rating);
      const mappedProps = mapDispatchToProps(mockDispatch)

      mappedProps.postRating(rating);
      expect(mockDispatch).toHaveBeenCalledWith(action);
    })
    // need delete and getRatings MDTP tests
  })
})
