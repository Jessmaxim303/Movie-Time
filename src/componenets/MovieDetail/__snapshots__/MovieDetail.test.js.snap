// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MovieDeatil should match a snapshot 1`] = `
<section
  className="movie-detail"
  key="21"
  style={
    Object {
      "background": "linear-gradient(to top, rgba(42, 42, 42, .75), rgba(42, 42, 42, 0)), url(https://image.tmdb.org/t/p/original//tCUcf3oNWMW8kwAj3WC6CvIN5ah.jpg) no-repeat center top",
      "backgroundSize": "cover",
    }
  }
>
  <div
    className="movie-detail__info"
  >
    <img
      alt="movie poster for Sonic the Hedgehog"
      className="movie-detail__poster"
      src="https://image.tmdb.org/t/p/original//aQvJ5WPzZgYVDrxLX4R6cLJCEaQ.jpg"
    />
    <div
      className="movie-detail__details"
    >
      <h2
        className="movie-detail__title"
      >
        Sonic the Hedgehog
      </h2>
      <p
        className="movie-detail__release-date"
      >
        Release Date: 2020-02-12
      </p>
      <p
        className="movie-detail__average-rating"
      >
        Average Rating: 7.00
      </p>
      <div
        className="movie-detail__user-rating"
      >
        <label
          for="currentRating"
        >
          User Rating: 
          -
        </label>
        <p
          className="movie-detail__rating-inactive"
        >
          Thank you for rating!
        </p>
      </div>
      <input
        className="movie-detail__rating-input"
        id="user-rating"
        max="10"
        min="1"
        name="currentRating"
        onChange={[Function]}
        type="range"
        value={0}
      />
      <div
        className="movie-detail__rating-buttons"
      >
        <button
          className="movie-detail__button movie-detail__delete-rating"
          onClick={[Function]}
        >
          Delete Rating
        </button>
        <button
          className="movie-detail__button movie-detail__add-rating"
          onClick={[Function]}
        >
          Add Rating
        </button>
      </div>
      <p
        className="movie-detail__overview"
      >
        Based on the global blockbuster videogame franchise from Sega, Sonic the Hedgehog tells the story of the worldâ€™s speediest hedgehog as he embraces his new home on Earth. In this live-action adventure comedy, Sonic and his new best friend team up to defend the planet from the evil genius Dr. Robotnik and his plans for world domination.
      </p>
    </div>
  </div>
</section>
`;
